atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStore.java

```diff
@@ -43,7 +52,24 @@
 /**
  * A PuncherStore implemented as a table in the KeyValueService.
  *
- * @author jweel
+ * This utilises dynamic columns to remove the necessity of range scans.
+ *
+ * The schema is:
+ *
+ * row: time rounded to 14 days
+ * column: rest of the timestamp, with inverted bits to ensure reverse range scans
+ * value: Atlas timestamp
+ *
+ * In order to look up a value for a give time, truncate in the same way and range scan within the row
+ * for the first cell beneath the provided time.
+ *
+ * Selecting the wall clock timestamp for a given Atlas timestamp still requires a full table scan,
+ * but should not be used in production code.
+ *
+ * It is expected that each row contains (with default setting) (1440 * 14 = 20k cells).
+ *
+ * Note one caveat of this implementation - as written, the get method only works if it is in the last
+ * 420 days, and defaults to Long.MIN_VALUE otherwise.
```

Comment:
I'm a bit torn about this, but I can see why it could be beneficial, and the scenario should quickly correct itself (i.e. the puncher should punch something, which should mean that in an hour or so in the worst case future gets should work). 

The consequences of giving a pessimistic answer to get is that sweep will stop, but it's unlikely and as mentioned self-correcting.

Reply:
exactly